File: 2025/Dart/Day12/last.dart
import 'dart:io';

int regionsCanFitPresents() {
  final file = File('input.txt');
  if (!file.existsSync()) {
    throw Exception('File input.txt not found');
  }

  final lines = file.readAsLinesSync();
  final buffer = StringBuffer();
  var readingPatterns = true;
  final patterns = <int>[];
  final regionLines = <String>[];

  for (final line in lines) {
    if (line.isEmpty) {
      if (readingPatterns && buffer.isNotEmpty) {
        patterns.add(buffer.toString().split('#').length - 1);
        buffer.clear();
      }
      continue;
    }

    if (readingPatterns) {
      if (line.contains(': ')) {
        readingPatterns = false;
        if (buffer.isNotEmpty) {
          patterns.add(buffer.toString().split('#').length - 1);
          buffer.clear();
        }
        regionLines.add(line);
      } else {
        if (buffer.isNotEmpty) {
          buffer.write('\n');
        }
        buffer.write(line);
      }
    } else {
      regionLines.add(line);
    }
  }

  if (readingPatterns && buffer.isNotEmpty) {
    patterns.add(buffer.toString().split('#').length - 1);
  }

  var regions = 0;

  for (final line in regionLines) {
    final colonIndex = line.indexOf(': ');
    if (colonIndex == -1) {
      continue;
    }

    final areaStr = line.substring(0, colonIndex);
    final numsStr = line.substring(colonIndex + 2);

    final xIndex = areaStr.indexOf('x');
    if (xIndex == -1) {
      continue;
    }

    final width = int.tryParse(areaStr.substring(0, xIndex));
    final height = int.tryParse(areaStr.substring(xIndex + 1));
    if (width == null || height == null) {
      continue;
    }

    final area = width * height;

    final fields = numsStr.split(RegExp(r'\s+')).where((s) => s.isNotEmpty);
    var size = 0;

    var i = 0;
    for (final field in fields) {
      if (i >= patterns.length) {
        break;
      }
      final num = int.tryParse(field);
      if (num != null) {
        size += patterns[i] * num;
      }
      i++;
    }

    if (size > area) {
      continue;
    }

    if (size * 1.2 < area) {
      regions++;
    }
  }

  print('Regions: $regions');
  return regions;
}

void main() {
  regionsCanFitPresents();
}

File: 2025/Go/Day12/last.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func RegionsCanFitPresents() int {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var builder strings.Builder
	readingPatterns := true
	var patterns []int
	var regionLines []string

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			if readingPatterns && builder.Len() > 0 {
				patterns = append(patterns, strings.Count(builder.String(), "#"))
				builder.Reset()
			}
			continue
		}

		if readingPatterns {
			if strings.Contains(line, ": ") {
				readingPatterns = false
				if builder.Len() > 0 {
					patterns = append(patterns, strings.Count(builder.String(), "#"))
					builder.Reset()
				}
				regionLines = append(regionLines, line)
			} else {
				if builder.Len() > 0 {
					builder.WriteString("\n")
				}
				builder.WriteString(line)
			}
		} else {
			regionLines = append(regionLines, line)
		}
	}

	if readingPatterns && builder.Len() > 0 {
		patterns = append(patterns, strings.Count(builder.String(), "#"))
	}

	var regions int = 0

	for _, line := range regionLines {
		colonIndex := strings.Index(line, ": ")
		if colonIndex == -1 {
			continue
		}

		areaStr := line[:colonIndex]
		numsStr := line[colonIndex+2:]

		xIndex := strings.Index(areaStr, "x")
		if xIndex == -1 {
			continue
		}

		width, err1 := strconv.Atoi(areaStr[:xIndex])
		height, err2 := strconv.Atoi(areaStr[xIndex+1:])
		if err1 != nil || err2 != nil {
			continue
		}

		area := width * height

		fields := strings.Fields(numsStr)
		size := 0

		for i, field := range fields {
			if i >= len(patterns) {
				break
			}
			if num, err := strconv.Atoi(field); err == nil {
				size += patterns[i] * num
			}
		}

		if size > area {
			continue
		}

		if float64(size)*(1.2) < float64(area) {
			regions++
		}
	}
	fmt.Printf("Regions: %d", regions)
	return regions
}

// func main() {
// 	RegionsCanFitPresents()

// }
